src/vm_interrupts.rs

```diff
@@ -25,11 +27,56 @@ enum BindStatus {
     Unbound,
 }
 
+const ALLOW_LIST_ENTRIES: usize = MAX_INTERRUPT_IDS / 64;
+
+// Bitmap tracking the per-vCPU allowed external interrupts.
+struct AllowList {
+    bits: ArrayVec<u64, ALLOW_LIST_ENTRIES>,
+    num_ids: usize,
+}
+
+impl AllowList {
+    fn new(num_ids: usize) -> Self {
+        let mut bits = ArrayVec::new();
+        let entries = (num_ids + 63) / 64;
+        for _ in 0..entries {
+            bits.push(0);
+        }
+
+        Self { bits, num_ids }
+    }
+
+    fn allow_id(&mut self, id: usize) -> Result<()> {
+        if id == 0 || id >= self.num_ids {
+            return Err(Error::InvalidInterruptId(id));
+        }
+        self.bits[id / 64] |= 1 << (id % 64);
+        Ok(())
+    }
+
+    fn allow_all(&mut self) {
+        self.bits.iter_mut().for_each(|i| *i = !0u64);
+    }
+
+    fn deny_id(&mut self, id: usize) -> Result<()> {
+        if id == 0 || id >= self.num_ids {
+            return Err(Error::InvalidInterruptId(id));
+        }
+        self.bits[id / 64] &= !(1 << (id % 64));
+        Ok(())
+    }
+
+    fn deny_all(&mut self) {
+        self.bits.iter_mut().for_each(|i| *i = 0u64);
+    }
+}
+
 /// Virtual external interrupt state for a vCPU.
 pub struct VmCpuExtInterrupts {
     bind_status: BindStatus,
     imsic_location: ImsicLocation,
     sw_file: SwFile,
+    allow_list: AllowList,
```

Comment:
naming nit: allowed_ids? something that indicates what is held as opposed to how it's implemented.

Reply:
Sure, done.